### PR TITLE
feat: support Google Gemini

### LINKED ISSUE #1171: Google Gemini API support
### Search before asking

- [X] 在 [issues](https://github.com/yetone/openai-translator/issues) 中没有找到类似的内容。
I searched in the [issues](https://github.com/yetone/openai-translator/issues) and found nothing similar.

### feature

added Google Gemini API support

### 描述 Motivation

From the twitter:  https://x.com/sundarpichai/status/1734952757722001626.

> Today developers can start building with our first version of Gemini Pro through Google AI Studio at [http://ai.google.dev](http://ai.google.dev). 
> 
> Developers have a free quota and access to a full range of features including function calling, embeddings, semantic retrieval, custom knowledge grounding, chat functionality and more. It supports 38 languages across 180+ countries. Gemini Ultra is coming early next year. We’re excited to see what you build! https://blog.google/technology/ai/gemini-api-developers-cloud/.

### Solution

_No response_

### 还有其他内容吗？Anything else?

_No response_

### 你是否愿意提交一份 PR？Are you willing to submit a PR?

- [ ] 我愿意提供 PR! I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by sinchang**
Closes #1171 

**How to test**

- Go to the[ Google AI Studio](https://makersuite.google.com/app/apikey) to get your API Key.

**Screenshot**

image

image

### CODE REVIEW

File: src/common/engines/gemini.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                    parts:
+                        req.assistantPrompts?.map((prompt) => ({
+                            text: prompt,
+                        })) ?? [],
+                },
+            ],
+            safetySettings: SAFETY_SETTINGS,
+        }
+
+        const response = await fetch(url, {
```
Comment (yetone) [inline src/common/engines/gemini.ts:72]:
```suggestion
	const fetcher = getUniversalFetch()
        const response = await fetcher(url, {
```

File: src/common/engines/gemini.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            method: 'POST',
+            headers,
+            body: JSON.stringify(body),
+        })
+
+        const json = await response.json()
```
Comment (yetone) [inline src/common/engines/gemini.ts:78]: ‌‌Does Gemini not support streaming?

**Comment by yetone** [comment]
‌‌‌You may need to update your code so that it can pass the CI.

File: src/common/engines/gemini.ts (hunk 2 above)
Comment (yetone) [reply src/common/engines/gemini.ts:78]: ‌‌‌‌I just looked at the document, Gemini API does support streaming, so this should be changed to use the streaming API.

https://ai.google.dev/tutorials/rest_quickstart#stream_generate_content

**Comment by yetone** [comment]
由于这个 PR 还需要一些工作才能合并，为了加快这个需求 release 的时间，同时为了不丢失你的贡献，所以我基于你的这个 branch checkout 出了一个新的 branch，所以关闭此 PR，在新的 PR 中继续开发：https://github.com/openai-translator/openai-translator/pull/1189

**Comment by sinchang** [comment]
> Since this PR still needs some work before it can be merged, in order to speed up the release time of this requirement, and also not to lose your contribution, I have checked out a new branch based on your branch. Therefore, this PR is closed and development will continue in the new PR：#1189

thank you

**Pull request closed without merging**